xc7/fasm2bels/cmt_models.py

```diff
@@ -0,0 +1,145 @@
+from .verilog_modeling import Bel, Site
+
+# =============================================================================
+
+def get_pll_site(db, grid, tile, site):
+    """ Return the prjxray.tile.Site object for the given PLL site. """
+    gridinfo = grid.gridinfo_at_tilename(tile)
+    tile_type = db.get_tile_type(gridinfo.tile_type)
+
+    sites = list(tile_type.get_instance_sites(gridinfo))
+    assert len(sites) == 1, sites
+
+    return sites[0]
+
+
+def decode_multi_bit_feature(features, target_feature):
+    """
+    Decodes a "multi-bit" fasm feature. If not present returns 0.
+    """
+    value = 0
+
+    for f in features:
+        last_part = f.feature.split('.')[-1]
+        if last_part.startswith(target_feature):
+            for canon_f in fasm.canonical_features(f):
+                if canon_f.start is None:
+                    value |= 1
+                else:
+                    value |= (1 << canon_f.start)
+
+    return value
+
+
+def process_pll(conn, top, tile_name, features):
+    """
+    Processes the PLL site
+    """
+
+    # Filter only PLL related features
+    pll_features = [f for f in features if 'PLLE2.' in f.feature]
+    if len(pll_features) == 0:
+        return
+
+    # Create the site
+    site = Site(pll_features, get_pll_site(top.db, top.grid, tile=tile_name, site='PLLE2_ADV'))
+
+    # If the PLL is not used then skip the rest
+    if not site.has_feature("IN_USE"):
+        return
+
+    # Create the PLLE2_ADV bel and add its ports
+    pll = Bel('PLLE2_ADV')
+
+    for i in range(6):
+        site.add_sink(pll, 'DADDR[{}]'.format(i), 'DADDR{}'.format(i))
+
+    for i in range(15):
+        site.add_sink(pll, 'DI[{}]'.format(i), 'DI{}'.format(i))
+
+    site.add_sink(pll, 'DCLK', 'DCLK')
+    site.add_sink(pll, 'DEN', 'DEN')
+    site.add_sink(pll, 'DWE', 'DWE')
+    site.add_sink(pll, 'CLKIN1', 'CLKIN1')
+    site.add_sink(pll, 'CLKIN2', 'CLKIN2')
+    site.add_sink(pll, 'CLKINSEL', 'CLKINSEL')
+    site.add_sink(pll, 'CLKFBIN', 'CLKFBIN')
+    site.add_sink(pll, 'RST', 'RST')
+    site.add_sink(pll, 'PWRDWN', 'PWRDWN')
+
+    site.add_source(pll, 'DRDY', 'DRDY')
+    site.add_source(pll, 'LOCKED', 'LOCKED')
+
+    for i in range(15):
+        site.add_source(pll, 'DO[{}]'.format(i), 'DO{}'.format(i))
+
+    # Process clock outputs
+    clkouts = ['FBOUT'] + ['OUT{}'.format(i) for i in range(6)]
+
+    for clkout in clkouts:
+        if site.has_feature('CLK{}_CLKOUT1_OUTPUT_ENABLE'.format(clkout)):
+
+            # Add output source
+            site.add_source(pll, 'CLK'+clkout, 'CLK'+clkout)
+
+            # Calculate the divider and duty cycle
+            high_time = decode_multi_bit_feature(features, 'CLK{}_CLKOUT1_HIGH_TIME'.format(clkout))
+            low_time  = decode_multi_bit_feature(features, 'CLK{}_CLKOUT1_LOW_TIME'.format(clkout))
+
+            divider = high_time + low_time
+            duty = high_time / (low_time + high_time)
```

Comment:
This needs to handle the EDGE feature, https://github.com/SymbiFlow/prjxray-db/blob/master/artix7/segbits_cmt_top_l_upper_t.db#L27
The EDGE enables half-step divider changes, see XAPP888.

Reply:
Fixed that. Another thing is that DivReg also contains an EDGE bit but there seems no control attribute on the PLLE2_ADV that can control it.

Reply:
When DIVCLK_DIVIDE is odd, DIVREG.EDGE is set I believe?